### ISSUE TITLE
dependsOn and required: true

### ISSUE THREAD

**Issue by morenoh149** [keystonejs/keystone#1011]
Let's suppose you're building a blog with keystonejs, a la https://github.com/JedWatson/keystone-demo/blob/master/models/posts.js#L14
Let's say your client has the requirement `All published posts should have a publishedDate`. Keystone has a convenience facility [dependsOn](http://keystonejs.com/docs/database/#fields) See `conditional Fields`. @webteckie proposed that combining `dependsOn` with `required: true` would properly enforce the above requirement. It does not. Instead the `require` forces all posts to then have a `publishedDate` - when most posts start out as drafts.

I was able to solve this with a mongoose pre-save hook https://gist.github.com/morenoh149/463fa1e717e67a6df228#file-post-js-L69

@keystonejs/contributors is this a bug? should keystone Lists support this usecase?

**Comment by webteckie** [comment]
The short version of @morenoh149 use case is this:
```
	state: { type: Types.Select, options: 'draft, published, archived', default: 'draft' },
	publishedDate: { type: Types.Date, dependsOn: {state: 'published'}, required: true },
```

That seems logical and it should work as expected, IMO.  That is, the dependsOn attribute condition should take precedence over the required attribute setting.  If dependsOn===false then required should not be enforced.

**Comment by JohnnyEstilles** [comment]
@morenoh149, like you, I was likewise skeptical when I first read about this feature. It felt counter-intuitive to conditionally `require` a field. Especially when tryng to make sense of it in terms of the *blog post state/publish date* use-case @webteckie presents above which, as you mention, can easily be enforced using pre-save hooks.

While I believe this is a very useful feature, I think the use of the  *blog post state/publish date* paradigm doesn't quite make the case for it.There are many other use-cases in which the concept of a conditionally required field is not only called for, but essential.

Take this scenario, for example: a field to capture a person's marital status, along with a conditional field for the name of that person's spouse, which is only required if the person is married.

```js
List.add({
    maritalStatus: { type: Types.Select, options: 'single, married, divorced, widowed', default: 'single' },
    spouse: { type: Types.Name, dependsOn: {maritalStatus: 'married'}, required: true }
});
```

Or maybe a scenario in which there you have a non-exhaustive list, represented by a `Types.Select` field which includes an `other` option to cover any options not listed in the select field, accompanied by a `String` field that is only required with `other` is selected.

```js
List.add({
    selectPetType: { type: Types.Select, options: 'dog, cat, tarantula, boa constrictor, skunk, miniature donkey, pygmy goat, other' },
    otherPetType: { type: Types.Name, dependsOn: {selectPetType: 'other'}, required: true }
});

List.schema.virtual('petType').get(function() {
	return this.selectPetType === 'other' ? this.otherPetType : this.selectPetType;
});
```

> Should keystone Lists support this usecase?

IMHO ... Definitely! I'm with @webteckie on this one! :-)

**Comment by webteckie** [comment]
@JohnnyEstilles FWIW, if I was implementing the requirements @morenoh149 had (force a publishDate if the user selected a 'publish' state for the post) then I would have certainly use the built-in solution as opposed to writing unnecessary logic.  However, if there's other pre-save processing needed then certainly ☺

**Comment by JohnnyEstilles** [comment]
@webteckie, I understand it's a matter of choice. I definitely agree with you. My point since the *blog post state/publish date* example can technically be solved multiple ways others may not be able to appreciate the value of what you were proposing.

I wasn't in any way trying to discourage us from implementing it. I'm all for it. :-)

**Comment by jossmac** [comment]
Reviving this issue. It seems that the case for this feature is solid.

Is anyone interested in having a go at a PR?

Ping @JedWatson

**Comment by morenoh149** [comment]
I am still interested in implementing this. @jossmac could you link to the file that would have to change for this to work?

**Comment by jossmac** [comment]
@morenoh149 I wasn't involved in the initially implementation, sorry. @JedWatson could you please point us in the right direction?

**Comment by snowkeeper** [comment]
I was looking through the codebase and went ahead and pushed  PR #2200 for you to review. 

For general list use I updated https://github.com/snowkeeper/keystone/blob/dependsOn-require/fields/types/Type.js#L68-L76 

I change `required` to a function if `dependsOn` is present
```
// if dependsOn and required, set required to a function for validation
	if(this.options.dependsOn && this.options.required === true) {
		var opts = this.options;
		this.options.required = function() {
			// `this` refers to the validating document
			debug('validate dependsOn required', evalDependsOn(opts.dependsOn, this.toObject()));
			return evalDependsOn(opts.dependsOn, this.toObject());
		}
	}
```

For the Admin UI I had to update https://github.com/snowkeeper/keystone/blob/dependsOn-require/lib/updateHandler.js#L272-L283
```
// validate required fields, unless they've already been invalidated by field-specific behaviour
		if (!invalidated && options.required[field.path] && !field.inputIsValid(data, true, this.item) && !validationErrors[field.path]) {
			// console.log('Field ' + field.path + ' is required, but not provided.');
			if(!field.dependsOn || evalDependsOn(field.dependsOn, data)) {
				debug('validate dependsOn required', field.dependsOn);
				// field is required only if dependsOn is true
				message = options.requiredMessages[field.path] || field.options.requiredMessage || field.label + ' is required';
				addValidationError(field.path, message);
				invalidated = true;	
			}
			
		}
```

**Comment by JedWatson** [comment]
Closed by #2200
